crates/shared/src/signature_validator/simulation.rs

```diff
@@ -94,6 +92,120 @@ impl Validator {
         tracing::trace!(?check, ?result, "simulated signature");
         Ok(Simulation { gas_used: result? })
     }
+
+    fn web3(&self) -> &Web3 {
+        &self.web3
+    }
+}
+
+#[async_trait::async_trait]
+impl Simulator for ZkSyncValidator {
+    #[instrument(skip_all, fields(interactions_len = check.interactions.len()))]
+    async fn simulate(
+        &self,
+        check: &SignatureCheck,
+    ) -> Result<Simulation, SignatureValidationError> {
+        // We simulate the signature verification from the Settlement contract's
+        // context. This allows us to check:
+        // 1. How the pre-interactions would behave as part of the settlement
+        // 2. Simulate the actual `isValidSignature` calls that would happen as part of
+        //    a settlement
+        let validate_call = self.signatures.methods().validate(
+            (self.settlement, self.vault_relayer),
+            check.signer,
+            Bytes(check.hash),
+            Bytes(check.signature.clone()),
+            check
+                .interactions
+                .iter()
+                .map(|i| (i.target, i.value, Bytes(i.call_data.clone())))
+                .collect(),
+        );
+        let random_account = random_account();
+        let storage_accessible = contracts::StorageAccessible::at(self.web3(), self.settlement);
```

Comment:
ditto: if Settlement exposes the `simulate_delegatecall` then I think it would be cleaner to use this contract instance directly.